### DISCUSSION TITLE
How to delete an item from inventory?

### DISCUSSION THREAD

**Discussion by Maklock** [Pryaxis/TShock#2107] [Q&A]
I'm learning how to make plugins right now,
I can actually change an items ID, Stack, Prefix, etc. but only server-side,
it doesn't show client-side, so I won't see the change until I relog to the server.
I know I'm supposed to use packets to achieve this, but after trying like 7 times I thought I could ask here.
How can I send the info from an inventory slot to the player's client-side?
(I'm using SSC)

**Comment by hakusaro** [comment]
If you're trying to play with the TShock SSC system (as opposed to just turning SSC on and using it manually) you need to update the TShock database and then send message 5 (player inventory slot) back to the client with the updated data. You can see an example here: https://github.com/Pryaxis/TShock/blob/general-devel/TShockAPI/PlayerData.cs#L417

The parameters for the packet are here, but you can use SendData in the same way that call uses it to achieve the expected result as opposed to crafting packets yourself and sending them over the wire. The precise packet is this: https://tshock.readme.io/docs/multiplayer-packet-structure#player-inventory-slot-5

When SSC is enabled on the server, the clients will take updates from the server as the true state and update their records according to the messages sent from the server. The reason why you have to relog if you've been playing with the database is that without sending the sync data, the client has no idea that the inventory has changed. You can either leverage the existing SSC system to update a player's inventory and have TShock "restore" the data, or you can reimplement things by hand (just depends on what you're going for).
GitHub <notifications@github.com> wrote:
“I'm learning how to make plugins right now,
 I can actually change an items ID, Stack, Prefix, etc. but only server-side,
 it doesn't show client-side, so I won't see the change until I relog to the server.
 I know I'm supposed to use packets to achieve this, but after trying like 7 times I thought I could ask here.
 How can I send the info from an inventory slot to the player's client-side?
 (I'm using SSC)”

**Reply by Maklock** [reply]
Thanks for all the info!
Got it working, but I have a question if you don't mind; seeing the table on readme.io,
you have player id, slot id, stack, prefix and netID, but then on the packet sent
`NetMessage.SendData(5, player.Index, -1, NetworkText.FromLiteral(Main.player[player.Index].inventory[k].Name), player.Index, slot, (float)Main.player[player.Index].inventory[k].prefix);`
You have: 5 (the packet, I get that), player index, -1 (?), item name, player index (again?), slot and prefix.
How do I get from the info on the table to that line?

**Reply by bartico6** [reply]
In order:
1. Packet Id (`5`)
2. Who it's sent to (`player.Index`)
3. Who it's *NOT* sent to (`-1`) - this is used when "who it's sent to" is set to -1, meaning "everyone" - you do this when telling everyone about someone's doings, the client who did the action doesn't need to be told about it. When argument 2, receiver, is not -1, this is unused.
4. Text (`NetworkText.FromLiteral(Main.player[player.Index].inventory[k].Name)`) - for packet 5 specifically this should be null, as it's never serialized into the packet in the current game version.
5. Whose inventory is being sent (`player.Index`) - **this is different than the receiver**.
6. Which slot is being sent (`slot`)
7. What prefix the item has (`(float)Main.player[player.Index].inventory[k].prefix`)

**Reply by Maklock** [reply]
That all makes sense now.
But I'm still wondering, how do you know that those are the values that should go on SendData(), like from the structure the Packet ID 5 has a stack value too, which is not used on the method call, so how do I know what to use and in what order? (I mean in general, like if I wanted to send a World Info packet (7) por example).

**Reply by bartico6** [reply]
Decompile the game and look at `SendData` in `NetMessage.cs`

Each case is a packet, and what's inside the switch case is the structure and what it takes as inputs.

Since the packets you mentioned (5 and 7) are too long to fit in one screen, I'll use character info (4) as an example:
![SendData(4)](https://i.quake.tf/1596417812-57062872.png)
"number", "number2", (...), "number7" are all parameters that go after "text" in SendData:
`public static void SendData(int msgType, int remoteClient = -1, int ignoreClient = -1, NetworkText text = null, int number = 0, float number2 = 0f, float number3 = 0f, float number4 = 0f, int number5 = 0, int number6 = 0, int number7 = 0)`
So, to recap:
- msgType is packet type, 
- remoteClient is recipient, 
- ignoreClient is who doesn't get the packet (if broadcasting through remoteClient = -1), 
and everything else (text + numbers) are parameters used for serialization.

**Reply by Maklock** [reply]
Are there any guides or resources on how to decompile the game?
Also, is that allowed by ReLogic?

**Comment by hakusaro** [comment]
The Pryaxis organization on GitHub has a repo for downloading and decompiling the game — just open it in Avalonia/ILSpy and it’ll be done in a jiffy.
GitHub <notifications@github.com> wrote:
“Are there any guides or resources on how to decompile the game?

 Also, is that allowed by ReLogic?”

**Reply by Maklock** [reply]
Nice! Thank you :)

**Reply by Maklock** [reply]
Is there something like that for the client-side as well?
I tried to decompile Terraria.exe with ILSpy but with no luck,
references don't load properly for some reason,
would be nice to have both sides decompiled.

**Reply by bartico6** [reply]
Both the client and server binary decompile just fine for me - try opening the game binary directly out of the game install folder, so it has the binaries it needs locally.
Also do you have XNA installed?

**Reply by Maklock** [reply]
I do have XNA installed, that one loads fine.
This is also my first time trying to decompile something, so I might be doing stuff wrong.
I open Terraria.exe on ILSpy, then right click Terraria and select "Load dependencies".

I get the following warning:

`Warning: Some assembly references could not be resolved automatically. This might lead to incorrect decompilation of some parts, for ex. property getter/setter access. To get optimal decompilation results, please manually add the missing references to the list of loaded assemblies.`

The references missing are Ionic.Zip.CF, RailSDK.Net, ReLogic and Steamworks.NET, so all of the .dll from the game's folder.
I tried opening those from ILSpy (or drag/dropping them to the list) without luck, I get the following exception:

```
ICSharpCode.Decompiler.Metadata.PEFileNotSupportedException: PE file does not contain any managed metadata.
   en ICSharpCode.Decompiler.Metadata.PEFile..ctor(String fileName, PEReader reader, MetadataReaderOptions metadataOptions)
   en ICSharpCode.ILSpy.LoadedAssembly.LoadAssembly(Object state)
   en System.Threading.Tasks.Task`1.InnerInvoke()
   en System.Threading.Tasks.Task.Execute()
```

I'm using the last version of ILSpy, I have visual studio 2019 installed and XNA 4.0, I also have the Core .NET 3.1 SDK installed.
ILSpy is set to C# 8.0. I tried for hours yesterday to get it working, but couldn't.. do you have any idea of what could be wrong?

**Reply by bartico6** [reply]
It's not an error: Just a warning.
The file you dragged in is ReLogic.Native.dll which is **not** ReLogic.dll.

Just decompile without loading dependencies, **it'll work**

**Reply by bartico6** [reply]
For background, the game has its dependencies **packed** as resources inside the game binary:
![Resources](https://i.quake.tf/1596567522-43767556.png)

The game will, at startup, unpack them in memory and load them, so you don't encounter errors.
ReLogic.Native.dll is not packed, just like CSteamworks because they're **native** binaries and they're not loaded into memory the same way.

You *can* unpack the .res files, rename them to .DLL without the Terraria.<...> part and try loading deps again but it won't yield much gain. Decompiling without manually loading resources **should** work. At least it does for me using dnSpy.

**Reply by Maklock** [reply]
Right, I tried saving code without loading dependencies, the project creates just fine.
It won't recompile though, seems like some methods have issues caused by missing references,
I see a lot of comments like

`//IL_0001: Unknown result type (might be due to invalid IL or missing references)`

Especially on the Net files, probably steamworks reference was missing and thus it couldn't generate proper code. So stuff like NetClientSocialModule.cs is just a mess.
But stuff like this is expected from decompiled code I guess.
I'll try to fix the files that have errors and hopefully it will compile just fine.

**Reply by bartico6** [reply]
You're not supposed to recompile the server this way.
